sqlccl: rewrite backup/restore implementation

danhhz opened the pull request:
This commit is a cherry-pick from the backup branch and essentially a
rewrite of everything in Backup and Restore, it's probably easiest to
review it as a rewrite. Sorry!

Notably, this commit adds support for backup and restore of interleaved
tables as well as pluggable storage for the backup (s3/gcs/azure).

Because of the order we're cherry-picking from the backup branch, this
does not yet support incremental backup.

Interleaved tables complicate restore in a few ways. First, our
(upcoming) Import and WriteBatch primitives require that the keyrange
being written to is empty. This means that we can't restore one table in
an interleave heirarchy and then another because the keyrange won't be
empty for the second table. This means we have to plumb the ability to
restore a set of tables at once.

Second, the additional complexity stressed various edge cases which
caused the existing tests to find some new bugs, most notably validation
that the timestamps for the incremental backups being used to restore
line up. Example: a full backup [0,A) and two incrementals [A,B),
[B,C). This property has to be true for every keyrange that's being
restored.

Finally, we weren't adjusting the interleave references in the
TableDescriptors being written. This commit now validates that all
interleave parents of any index are being restored (a restriction that
could be lifted with stub TableDescriptors), and delete references to
any tables not restored. This does mean that once we switch to Import,
later restoring one of these latter tables will be impossible; this
could be fixed by going back to this Put implementation.

Other changes:
- Move BackupRangeDescriptor to be a nested type
- One start time for all ranges in a backup. It's possible that we want
  to make this more general again at some point, but for now it's not
  worth it (supporting an incremental backup on top of two
  non-overlapping backups).
- Use ExportStroage in BackupDescriptor to abstract the target storage
  (s3/gcs/azure).).
- Remove IntersectHalfOpen and MakeRekeyMVCCKeyValFunc, which are no
  longer used.
- Move AllRangeDescriptors to tests and unexport.

Reviewers:
- We're switching from the ExportKeys/Ingest/Import used in the branch
  to Export/Import/Load, respectively.
- Note that is is being submitted before the Import and WriteBatch
  commands. Their functionality is purely performance optimizations
  (except for startTime != 0 in Export which powers incremental backup),
  so they're replaced here by kv ops. I haven't really updated any
  comments, so there's a bunch of references to them. Seems okay to me
  as an interim state, but I'll fix it up if anyone feels strongly.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13276)
<!-- Reviewable:end -->

danhhz committed:
sqlccl: rewrite backup/restore implementation

This commit is a cherry-pick from the backup branch and essentially a
rewrite of everything in Backup and Restore, it's probably easiest to
review it as a rewrite. Sorry!

Notably, this commit adds support for backup and restore of interleaved
tables as well as pluggable storage for the backup (s3/gcs/azure).

Because of the order we're cherry-picking from the backup branch, this
does not yet support incremental backup.

Interleaved tables complicate restore in a few ways. First, our
(upcoming) Import and WriteBatch primitives require that the keyrange
being written to is empty. This means that we can't restore one table in
an interleave heirarchy and then another because the keyrange won't be
empty for the second table. This means we have to plumb the ability to
restore a set of tables at once.

Second, the additional complexity stressed various edge cases which
caused the existing tests to find some new bugs, most notably validation
that the timestamps for the incremental backups being used to restore
line up. Example: a full backup [0,A) and two incrementals [A,B),
[B,C). This property has to be true for every keyrange that's being
restored.

Finally, we weren't adjusting the interleave references in the
TableDescriptors being written. This commit now validates that all
interleave parents of any index are being restored (a restriction that
could be lifted with stub TableDescriptors), and delete references to
any tables not restored. This does mean that once we switch to Import,
later restoring one of these latter tables will be impossible; this
could be fixed by going back to this Put implementation.

Other changes:
- Move BackupRangeDescriptor to be a nested type
- One start time for all ranges in a backup. It's possible that we want
  to make this more general again at some point, but for now it's not
  worth it (supporting an incremental backup on top of two
  non-overlapping backups).
- Use ExportStroage in BackupDescriptor to abstract the target storage
  (s3/gcs/azure).).
- Remove IntersectHalfOpen and MakeRekeyMVCCKeyValFunc, which are no
  longer used.
- Move AllRangeDescriptors to tests and unexport.

Reviewers:
- We're switching from the ExportKeys/Ingest/Import used in the branch
  to Export/Import/Load, respectively.
- Note that is is being submitted before the Import and WriteBatch
  commands. Their functionality is purely performance optimizations
  (except for startTime != 0 in Export which powers incremental backup),
  so they're replaced here by kv ops. I haven't really updated any
  comments, so there's a bunch of references to them. Seems okay to me
  as an interim state, but I'll fix it up if anyone feels strongly.

danhhz commented:
@a-robinson @RaduBerinde My first priority this week is the partitioning rfc, so there's no hurry on this.

danhhz commented:
Okay, @mjibson is going to start back on some backup/restore stuff this week and we'll need this in before that happens. @a-robinson or @RaduBerinde, think you'll get a chance to look at this? I realize it's huge, so happy to help however I can

a-robinson commented:
I'll bite the bullet. Hopefully I'll get through it all by early-mid afternoon.

a-robinson commented:
Nice work on this, @danhhz!

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJaOCdikEg7PQaS-hi:-KcJaOCdikEg7PQaS-hj:bv0rgvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L63)):*
> ```Go
> 		return BackupDescriptor{}, err
> 	}
> 	return backupDesc, nil
> ```

Speaking as someone who knows nothing about what's in the backup descriptors, is any validation needed to make sure that it still makes sense or that it hasn't been corrupted?

---

*[pkg/ccl/sqlccl/backup.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJdXmqPtifcXJkUGGR:-KcJdXmqPtifcXJkUGGS:bwlz0pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L74)):*
> ```Go
> 	if len(paths) == 0 || len(paths) == 1 && paths[0] == "" {
> 		// Full backup.
> 		return hlc.Timestamp{}, nil
> ```

It'd perhaps be more readable to use hlc.ZeroTimestamp for this (and for endTime below)

---

*[pkg/ccl/sqlccl/backup.go, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJeH8gM56l1LBfNXjz:-KcJeH8hANSWHItKtPii:b-y0dvd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L119)):*
> ```Go
> func (r backupFileDescriptors) Swap(i, j int) { r[i], r[j] = r[j], r[i] }
> func (r backupFileDescriptors) Less(i, j int) bool {
> 	return r[i].StartKey.Compare(r[j].EndKey) == -1
> ```

Just to double check, it's impossible for a single key to span more than one range, right?

---

*[pkg/ccl/sqlccl/backup.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJh0Cc4UXG-FhM_D8-:-KcJh0Cc4UXG-FhM_D80:bz489cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L160)):*
> ```Go
> 	// TODO(dan): Tighten this up to only the data for the tables requested in
> 	// the BACKUP request.
> 	descriptorStartKey := roachpb.Key(keys.MakeTablePrefix(keys.DescriptorTableID))
> ```

I'm kind of impressed that the descriptor table is the only special table we need to back up. None of the other system tables are needed?

---

*[pkg/ccl/sqlccl/backup.go, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJliT8-hUphBgzslyx:-KcJliT8-hUphBgzslyy:b-uhfj9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L199)):*
> ```Go
> 		return BackupDescriptor{}, err
> 	}
> 	defer writer.Close()
> ```

Not knowing the implementation of the writer, is it safe to ignore the error from `Close()`? What is it doing that `Finish()` isn't?

---

*[pkg/ccl/sqlccl/backup.go, line 218 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJn8HWDqqmBEHpHfH4:-KcJn8HWDqqmBEHpHfH5:bkgg5kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L218)):*
> ```Go
> 	span roachpb.Span,
> 	exportStore storageccl.ExportStorage,
> ) (res BackupDescriptor_File, dataSize int64, retErr error) {
> ```

It's kind of bad form to name the return values but then separately specify the values in each return statement. You should just remove the names here (and comment above what the int64 represents)

---

*[pkg/ccl/sqlccl/backup.go, line 226 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmTZzK7V95h-vOP2W:-KcJmTZzK7V95h-vOP2X:bvzvpl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L226)):*
> ```Go
> 	}
> 	path := writer.LocalFile()
> 	defer writer.Close()
> ```

Same `Close()` safety question here

---

*[pkg/ccl/sqlccl/backup.go, line 228 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmXoasYPqKw6T4H0v:-KcJmXoasYPqKw6T4H0w:balhh9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L228)):*
> ```Go
> 	defer writer.Close()
> 
> 	sstWriter := engine.MakeRocksDBSstFileWriter()
> ```

Not your doing, but this should be `MakeRocksDBSSTFileWriter`

---

*[pkg/ccl/sqlccl/backup.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJnkRdpjkB8KaCxsQK:-KcJnkRfLpEFUY6_2Qmy:b-9yht35) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L234)):*
> ```Go
> 	}
> 	defer func() {
> 		if retErr != nil && sst != nil {
> ```

This logic scares me. You aren't actually setting `retErr` anywhere as far as I can tell, so why are you checking it here? And why would an error mean that we shouldn't try closing `sst` anyway?

---

*[pkg/ccl/sqlccl/backup.go, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmrVjOtMTwFLv4YpB:-KcJmrVk9vV_1PtbgT0X:b-67gxw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L236)):*
> ```Go
> 		if retErr != nil && sst != nil {
> 			if closeErr := sst.Close(); closeErr != nil {
> 				log.Warningf(ctx, "Could not close sst writer %s", path)
> ```

For whatever reason, our convention is to start log messages with lower case letters

---

*[pkg/ccl/sqlccl/backup.go, line 239 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmveiqzssYnglwejz:-KcJmveiqzssYnglwek-:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L239)):*
> ```Go
> 			}
> 			if delErr := os.Remove(path); delErr != nil {
> 				log.Warningf(ctx, "Could not clean up %s", path)
> ```

Ditto

---

*[pkg/ccl/sqlccl/backup.go, line 285 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJoNRIu3rpvHs-amhB:-KcJoNRIu3rpvHs-amhC:bx3t1ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L285)):*
> ```Go
> 		return BackupDescriptor_File{}, 0, err
> 	}
> 	sst = nil
> ```

If we fail after this statement, is it an issue that we'll be leaking the SST file?

---

*[pkg/ccl/sqlccl/backup.go, line 335 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJp4evcg2ZvPqvy_eH:-KcJp4ewMLTauOFoXQPA:b-ms9i3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L335)):*
> ```Go
> 	}
> 
> 	// Arrived at by tuning and watching the effect on BenchmarkRestore.
> ```

Did you record that tuning effort anywhere? If so, it might be interesting to link to for posterity

---

*[pkg/ccl/sqlccl/backup.go, line 392 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJptr4bHsfsO6tN7fe:-KcJptr4bHsfsO6tN7ff:b-cfwkbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L392)):*
> ```Go
> 		}
> 
> 		localPath, err := base.FetchFile(ctx, file.Path)
> ```

I sincerely hope that this file never needs to import the `base` package :)

---

*[pkg/ccl/sqlccl/backup.go, line 402 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJq7-b_m-H34I7J9ZE:-KcJq7-b_m-H34I7J9ZF:b-7y0j6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L402)):*
> ```Go
> 		}
> 		defer sst.Close()
> 		if err := sst.AddFile(localPath); err != nil {
> ```

I imagine that there is or else you wouldn't have done this, but is there any reason why we couldn't just add all the file paths to this sst reader at once rather than creating a separate sst reader for each file?

---

*[pkg/ccl/sqlccl/backup.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJrWN3ScKsn93ZwGfC:-KcJrWN4tgfHxUeKGhJ_:b-ifj2ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L541)):*
> ```Go
> 			indexSSTStartKey := roachpb.Key(sqlbase.MakeIndexKeyPrefix(table, index.ID))
> 			indexSSTEndKey := indexSSTStartKey.PrefixEnd()
> 			ie := importEntry{Span: roachpb.Span{
> ```

I'm a little surprised to see that not every importEntry has a type (and/or that not every backupEntry has it's `backup` field set). Maybe I'm just reading into it too much, or maybe  the type definition would benefit from a comment

---

*[pkg/ccl/sqlccl/backup.go, line 837 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJtS6GgeBQZR2UcCuy:-KcJtS6GgeBQZR2UcCuz:b-jxmiul) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L837)):*
> ```Go
> func userTablesAndDBsMatchingName(
> 	descs []sqlbase.Descriptor, name parser.TableName,
> ) (map[sqlbase.ID]*sqlbase.DatabaseDescriptor, []*sqlbase.TableDescriptor, error) {
> ```

It seems like it's time to add a comment clarifying what the return values are

---

*[pkg/ccl/sqlccl/backup.go, line 876 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJtjI53AG0h6nG7uRx:-KcJtjI53AG0h6nG7uRy:b37uxzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L876)):*
> ```Go
> 		base, err := storageccl.ExportStorageFromURI(ctx, basePath)
> 		if err != nil {
> 			return nil, errors.Wrapf(err, "failed to create export stoage handler from %q", basePath)
> ```

s/stoage/storage

---

*[pkg/ccl/sqlccl/backup.proto, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJuqAHxEph6q-xztYN:-KcJuqAHxEph6q-xztYO:bgxv3ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.proto#L37)):*
> ```protobuf
>     bytes end_key = 3 [(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.Key"];
>     uint32 crc = 4 [(gogoproto.customname) = "CRC"];
>     int32 node_id = 5 [(gogoproto.customname) = "NodeID",
> ```

I assume this is just forward-looking? It doesn't appear to be used anywhere.

---

*[pkg/ccl/sqlccl/backup.proto, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJv08l-phD4P5ZTlUd:-KcJv08l-phD4P5ZTlUe:b-msx2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.proto#L48)):*
> ```protobuf
>   repeated roachpb.Span spans = 3 [(gogoproto.nullable) = false];
>   repeated File files = 4 [(gogoproto.nullable) = false];
>   repeated sql.sqlbase.Descriptor sql = 5 [(gogoproto.nullable) = false, (gogoproto.customname) = "SQL"];
> ```

It'd probably be too annoying to change this now, but `sql` is a weirdly non-descriptive name for a field inside a SQL database's backup format.

---

*[pkg/ccl/sqlccl/backup_test.go, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJvyc7p5fT3WQPoqpm:-KcJvyc9BAUsJldiHlAi:bn1tv6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup_test.go#L183)):*
> ```Go
> }
> 
> func TestBackupRestoreS3(t *testing.T) {
> ```

It looks like this and the GCS test actually hit S3 and GCS for real? You might want to leave a comment on them indicating as such to help people looking into the occasional flake caused by them.

Is teamcity going to have keys to write to GCS/S3 or will these only run when a developer specifically sets the relevant env vars?

---

*[pkg/ccl/sqlccl/backup_test.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJvWCzIdLNWoMCCrdY:-KcJvWCzIdLNWoMCCrdZ:bgtdz82) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup_test.go#L186)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	s3Keys, err := s3gof3r.EnvKeys()
> ```

That may be the most annoying package name I've ever seen...

---

*[pkg/ccl/sqlccl/backup_test.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJx04K65-uI_BbN1rr:-KcJx04K65-uI_BbN1rs:b-828014) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup_test.go#L318)):*
> ```Go
> 		sqlDBRestore.QueryRow(`SELECT COUNT(*) FROM bench.bank`).Scan(&rowCount)
> 		if rowCount != numAccounts {
> 			t.Fatalf("expected %d rows but found %d", numAccounts, rowCount)
> ```

Do these really need to be `t.Fatalf` or would it be useful to switch them to `t.Errorf` and continue running the other checks?

---

*[pkg/sql/sqlbase/structured.go, line 333 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJxIAk4h7bSGSiI_m0:-KcJxIAk4h7bSGSiI_m1:bvgexkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/sql/sqlbase/structured.go#L333)):*
> ```Go
> // added in the mutations.
> func (desc *TableDescriptor) ForeachNonDropIndex(f func(*IndexDescriptor) error) error {
> 	if desc.IsPhysicalTable() {
> ```

Can non-physical tables even have indexes? Whoa

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-:-KcJxUc4jBvdVELiT0WS:b-dso1ja)*
<!-- Sent from Reviewable.io -->

danhhz committed:
storageccl: replace ExportFileWriter.Close with Cleanup

It is required to be idempotent and any errors should be logged (not
returned). This makes the common usage of this interface much nicer.

danhhz commented:
Thanks for the review! This is great stuff. RFAL

Unrelated to your comments, I changed up BackupDescriptor_File just a bit (moved StartKey and EndKey to be a Span) because I'd meant to before sending this out but forgot.

---

Review status: 0 of 9 files reviewed at latest revision, 24 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJaOCdikEg7PQaS-hi:-KcKTndmj6B0JLZBhcnT:b-xkawyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L63)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Speaking as someone who knows nothing about what's in the backup descriptors, is any validation needed to make sure that it still makes sense or that it hasn't been corrupted?
</blockquote></details>

Good idea. Added a TODO with some of the things we could check for.

---

*[pkg/ccl/sqlccl/backup.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJdXmqPtifcXJkUGGR:-KcKU-Na9ZQiGVnme8YN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L74)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It'd perhaps be more readable to use hlc.ZeroTimestamp for this (and for endTime below)
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJeH8gM56l1LBfNXjz:-KcKUgiz2bUnXUvYfdUn:bf0hipn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L119)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Just to double check, it's impossible for a single key to span more than one range, right?
</blockquote></details>

Correct. Also, this impl is weird now that you point it out. Make it more standard.

---

*[pkg/ccl/sqlccl/backup.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJh0Cc4UXG-FhM_D8-:-KcKVHait25UoqX1K3nn:b-q584n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L160)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'm kind of impressed that the descriptor table is the only special table we need to back up. None of the other system tables are needed?
</blockquote></details>

I have a TODO somewhere to add users to this list. Went ahead and added that one

---

*[pkg/ccl/sqlccl/backup.go, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJliT8-hUphBgzslyx:-KcKwHrgnjORtXpdT8Qc:bvts9c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L199)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Not knowing the implementation of the writer, is it safe to ignore the error from `Close()`? What is it doing that `Finish()` isn't?
</blockquote></details>

Reworked this a bit. See if it makes more sense now.

---

*[pkg/ccl/sqlccl/backup.go, line 218 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJn8HWDqqmBEHpHfH4:-KcL2pKxVqKMJN-kZtIm:bs2bxo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L218)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It's kind of bad form to name the return values but then separately specify the values in each return statement. You should just remove the names here (and comment above what the int64 represents)
</blockquote></details>

Done. This was here to facilitate my trickiness

---

*[pkg/ccl/sqlccl/backup.go, line 226 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmTZzK7V95h-vOP2W:-KcKwNG5iye10lu4xo6G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L226)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Same `Close()` safety question here
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 228 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmXoasYPqKw6T4H0v:-KcKivbqUBL_EtY68h9S:bmza28q) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L228)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Not your doing, but this should be `MakeRocksDBSSTFileWriter`
</blockquote></details>

Actually is my doing : - ) TODO'd

---

*[pkg/ccl/sqlccl/backup.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJnkRdpjkB8KaCxsQK:-KcKx0I809RmcUp-D_rx:b-h8jdax) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L234)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

This logic scares me. You aren't actually setting `retErr` anywhere as far as I can tell, so why are you checking it here? And why would an error mean that we shouldn't try closing `sst` anyway?
</blockquote></details>

I was being tricky. retErr was set by a return (including an early return), so the != nil check would tell you if it was an error case or not. It should be more obvious now

---

*[pkg/ccl/sqlccl/backup.go, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmrVjOtMTwFLv4YpB:-KcKWOc6WEB5i9FjDA_-:b423vfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L236)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

For whatever reason, our convention is to start log messages with lower case letters
</blockquote></details>

Done. Did a couple others, too

---

*[pkg/ccl/sqlccl/backup.go, line 239 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJmveiqzssYnglwejz:-KcKWP6Ls-nyEzq9SOQQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L239)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 285 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJoNRIu3rpvHs-amhB:-KcKxpb3yExob4I1LNPq:boxrupq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L285)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

If we fail after this statement, is it an issue that we'll be leaking the SST file?
</blockquote></details>

Fixed the leak and commented how the fix works.

---

*[pkg/ccl/sqlccl/backup.go, line 335 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJp4evcg2ZvPqvy_eH:-KcKWdnuxhJWsKY-OV22:b-8yugsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L335)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Did you record that tuning effort anywhere? If so, it might be interesting to link to for posterity
</blockquote></details>

Alas, no

---

*[pkg/ccl/sqlccl/backup.go, line 392 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJptr4bHsfsO6tN7fe:-KcKY-0f8S6l-ojsMoRA:bjexvyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L392)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I sincerely hope that this file never needs to import the `base` package :)
</blockquote></details>

Now it's `dir`. If anyone makes that package, I'm out of ideas : - )

---

*[pkg/ccl/sqlccl/backup.go, line 402 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJq7-b_m-H34I7J9ZE:-KcKYAQlXMi3Zu37h5W3:b-l4396t) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L402)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I imagine that there is or else you wouldn't have done this, but is there any reason why we couldn't just add all the file paths to this sst reader at once rather than creating a separate sst reader for each file?
</blockquote></details>

Added a comment

---

*[pkg/ccl/sqlccl/backup.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJrWN3ScKsn93ZwGfC:-KcKkewYetJ2LlcU7Dm-:b-yygwob) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L541)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'm a little surprised to see that not every importEntry has a type (and/or that not every backupEntry has it's `backup` field set). Maybe I'm just reading into it too much, or maybe  the type definition would benefit from a comment
</blockquote></details>

Yeah, I was abusing importEntry here for two things, but didn't really need to be. Done.

I can't tell if I fixed both halves of your question or just the first part. Take a look and let me know if I still need more comments

---

*[pkg/ccl/sqlccl/backup.go, line 837 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJtS6GgeBQZR2UcCuy:-KcLBtjYvvBq70iZKJHH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L837)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It seems like it's time to add a comment clarifying what the return values are
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 876 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJtjI53AG0h6nG7uRx:-KcKZBnfM1GOU39sIv8D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L876)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/stoage/storage
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.proto, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJuqAHxEph6q-xztYN:-KcKa2k-oHuJTY2IfaED:b-e0fdik) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.proto#L37)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I assume this is just forward-looking? It doesn't appear to be used anywhere.
</blockquote></details>

The opposite: deprecated. Thought I removed it but guess not.

---

*[pkg/ccl/sqlccl/backup.proto, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJv08l-phD4P5ZTlUd:-KcKa844I11d0GHc1x9W:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.proto#L48)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It'd probably be too annoying to change this now, but `sql` is a weirdly non-descriptive name for a field inside a SQL database's backup format.
</blockquote></details>

Done

---

*[pkg/ccl/sqlccl/backup_test.go, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJvyc7p5fT3WQPoqpm:-KcKm_o4R47J1fc6OJZZ:bpt3wrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup_test.go#L183)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It looks like this and the GCS test actually hit S3 and GCS for real? You might want to leave a comment on them indicating as such to help people looking into the occasional flake caused by them.

Is teamcity going to have keys to write to GCS/S3 or will these only run when a developer specifically sets the relevant env vars?
</blockquote></details>

Done. TBD on the second part.

---

*[pkg/ccl/sqlccl/backup_test.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJvWCzIdLNWoMCCrdY:-KcKmsPDcdf1z-ktW1c_:bejlpoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup_test.go#L186)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

That may be the most annoying package name I've ever seen...
</blockquote></details>

No kidding

---

*[pkg/ccl/sqlccl/backup_test.go, line 318 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJx04K65-uI_BbN1rr:-KcKn0hgPjZYGpSHhSuB:bp0yfac) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup_test.go#L318)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Do these really need to be `t.Fatalf` or would it be useful to switch them to `t.Errorf` and continue running the other checks?
</blockquote></details>

You are correct, sir. Done

---

*[pkg/sql/sqlbase/structured.go, line 333 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJxIAk4h7bSGSiI_m0:-KcKcBYTlsABf4HKijvq:bvytg5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/sql/sqlbase/structured.go#L333)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Can non-physical tables even have indexes? Whoa
</blockquote></details>

No idea. I was just cargo culting AllNonDropIndexes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-:-KcLDr7bOddKpVndc1IZ:b-b0laxr)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 2 of 8 files at r2, 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 27 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJliT8-hUphBgzslyx:-KcLsVoL3T0wgdUcc9eU:bbb0p1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L199)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Reworked this a bit. See if it makes more sense now.
</blockquote></details>

Yup, thanks!

---

*[pkg/ccl/sqlccl/backup.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJnkRdpjkB8KaCxsQK:-KcLss78MjSJ3DnQIx8A:b-2joulg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L234)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I was being tricky. retErr was set by a return (including an early return), so the != nil check would tell you if it was an error case or not. It should be more obvious now
</blockquote></details>

Soooo tricky :) Thanks for dumbing it down for me, I hadn't seen named return values used that way before

---

*[pkg/ccl/sqlccl/backup.go, line 335 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJp4evcg2ZvPqvy_eH:-KcLtKT7Y8mlZ-xXBdvm:b-ltg5t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L335)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Alas, no
</blockquote></details>

I wouldn't worry about rerunning it at this point, it doesn't seem worth the effort. Feel free to remove the TODO

---

*[pkg/ccl/sqlccl/backup.go, line 392 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJptr4bHsfsO6tN7fe:-KcLtXAYNeS3YulKeVoU:b-e3d1nd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L392)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Now it's `dir`. If anyone makes that package, I'm out of ideas : - )
</blockquote></details>

Heh, you didn't need to rename it, but I'll fight back hard if anyone ever tries to create a `dir` package

---

*[pkg/ccl/sqlccl/backup.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJrWN3ScKsn93ZwGfC:-KcLu33BnIVgkkIjJW8b:b-q67zfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L541)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Yeah, I was abusing importEntry here for two things, but didn't really need to be. Done.

I can't tell if I fixed both halves of your question or just the first part. Take a look and let me know if I still need more comments
</blockquote></details>

Nah, it's fine. I think I was being a little too nit-picky, I doubt that code really would have thrown anyone off.

---

*[pkg/ccl/sqlccl/backup.proto, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLufpHCA9OSTV2-umr:-KcLufpHCA9OSTV2-ums:bvs5y22) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/sqlccl/backup.proto#L33)):*
>     string path = 2;
>     uint32 crc = 3 [(gogoproto.customname) = "CRC"];

Any reason why these don't have `gogoproto.nullable` set to false? 

---

*[pkg/ccl/storageccl/export_storage.go, line 130 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLsOTgLt_2XSBLbSaV:-KcLsOTgLt_2XSBLbSaW:b-gk1q4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/storageccl/export_storage.go#L130)):*
> ```Go
> 	// LocalFile returns the path to a local path to which a caller should write.
> 	LocalFile() string
> 	// Finish indicates that no further writes to the local file are exepcted and
> ```

s/exepcted/expected

---

*[pkg/ccl/storageccl/export_storage.go, line 202 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLve3rMm3dYkjMi3UG:-KcLve3rMm3dYkjMi3UH:b-qotg3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/storageccl/export_storage.go#L202)):*
> ```Go
> func (l localFileStorageWriter) Cleanup() {
> 	if err := os.RemoveAll(l.LocalFile()); err != nil {
> 		log.Warningf(context.TODO(), "could not remove %s: %+v", l.LocalFile(), err)
> ```

nit: I might use %q instead of %s in case `l.LocalFile()` returns the empty string. And why do we need to print `err` with `%+v` rather than just `%s`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-:-KcLw0ftFQvanrkfWQ7f:b-uwkowd)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 7 of 9 files reviewed at latest revision, 27 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJnkRdpjkB8KaCxsQK:-KcO8XJNnWRfyxVgpP7I:bqcrx0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L234)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Soooo tricky :) Thanks for dumbing it down for me, I hadn't seen named return values used that way before
</blockquote></details>

Yeah, if you weren't seeing what was going on there, it was _too_ tricky

---

*[pkg/ccl/sqlccl/backup.go, line 335 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcJp4evcg2ZvPqvy_eH:-KcO9bQbZ_N_DZaQkqbx:bb74ce2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bffa839839382e15cdeb2b61c8ddf728b445b57e/pkg/ccl/sqlccl/backup.go#L335)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I wouldn't worry about rerunning it at this point, it doesn't seem worth the effort. Feel free to remove the TODO
</blockquote></details>

Sure

---

*[pkg/ccl/sqlccl/backup.proto, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLufpHCA9OSTV2-umr:-KcO99amgIKQtvp2dvOm:bh5z4fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/sqlccl/backup.proto#L33)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

>     string path = 2;
>     uint32 crc = 3 [(gogoproto.customname) = "CRC"];

Any reason why these don't have `gogoproto.nullable` set to false? 
</blockquote></details>

```ERROR: field File.Path is a native type and in proto3 syntax with nullable=false there exists conflicting implementations when encoding zero values--gogoroach_out```

¯\_(ツ)_/¯
```


---

*[pkg/ccl/storageccl/export_storage.go, line 130 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLsOTgLt_2XSBLbSaV:-KcO9JG6NtNgvWAPKxyN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/storageccl/export_storage.go#L130)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/exepcted/expected
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/export_storage.go, line 202 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLve3rMm3dYkjMi3UG:-KcO9cnCIBVaT08TpCaq:b6mpcam) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/storageccl/export_storage.go#L202)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

nit: I might use %q instead of %s in case `l.LocalFile()` returns the empty string. And why do we need to print `err` with `%+v` rather than just `%s`?
</blockquote></details>

Done on `%q` for the string.

In general, I always default to `%+v` for errors and don't think too hard about it. In the case that it is now (or later becomes) something that comes via `errors.Wrap`, you get a lot more info this way. I'm not really sure what the downside is, do you have any thoughts? Even if it's more expensive, I figure it's safe to assume error cases are rare.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Thanks, Alex!!!!!! You're a hero

tamird commented:
Review status: 7 of 9 files reviewed at latest revision, 27 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.proto, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLufpHCA9OSTV2-umr:-KcOLt2pQgWeenOU5-Fa:bzfp4aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/sqlccl/backup.proto#L33)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

```ERROR: field File.Path is a native type and in proto3 syntax with nullable=false there exists conflicting implementations when encoding zero values--gogoroach_out```

¯\_(ツ)_/¯
</blockquote></details>

and the CRC?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 7 of 9 files reviewed at latest revision, 27 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.proto, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276#-KcLufpHCA9OSTV2-umr:-KcOLxBlPsC24RmdCUw5:b76c30j) ([raw file](https://github.com/cockroachdb/cockroach/blob/354b8f7a1b0a9cc22153aea85d5f8b78c4802481/pkg/ccl/sqlccl/backup.proto#L33)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

and the CRC?
</blockquote></details>

oh, this whole file is proto3. ignore me!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13276)*
<!-- Sent from Reviewable.io -->
